smarts/core/controllers/trajectory_tracking_controller.py

```diff
@@ -531,25 +534,20 @@ def MPC(
         # gains are as [lateral_error,lateral_velocity,heading_error,yaw_rate].
         # Increasing the lateral_error weight can cause oscillations which can
         # be damped out by increasing yaw_rate weight.
-        Q_tilde = np.kron(np.eye(prediction_horizon), 0.1 * np.diag([154, 0, 14, 150]))
+        Q_tilde = np.kron(np.eye(prediction_horizon), 0.1 * np.diag([354, 0, 14, 250]))
         # R_tilde contain the steering input weight.
         R_tilde = np.kron(np.eye(prediction_horizon), np.eye(1))
         matrix_H = (np.transpose(matrix_C)).dot(Q_tilde).dot(matrix_C) + R_tilde
         matrix_F = (np.transpose(matrix_C)).dot(Q_tilde).dot(matrix_M)
 
         matrix_F1 = (np.transpose(matrix_C)).dot(Q_tilde).dot(matrix_T_tilde)
 
-        matrix_U = solvers.qp(
-            matrix(matrix_H),
-            matrix(
-                2
-                * np.matmul(
-                    matrix_F, np.array([[lateral_error], [0], [heading_error], [0]])
-                )
-                + 2 * np.reshape(matrix_F1, (-1, 1))
-            ),
-            matrix(np.eye(matrix_H.shape[0])),
-            matrix(np.ones((matrix_H.shape[0], 1))),
-        )
+        # This is the solution to the unconstrained optimization problem
+        # for the MPC.
+        unconstrained_optimal_solution = np.matmul(
+            np.linalg.inv(2 * matrix_H),
+            np.matmul(matrix_F, np.array([[lateral_error], [0], [heading_error], [0]]))
+            + np.reshape(matrix_F1, (-1, 1)),
+        )[0][0]
```

Comment:
that's it? cool

Reply:
Thanks, Yes, basically, the MPC formulation results in the following QP:

min (1/2)x^T*H*x+F^T*X, subject to A*X<=B
  x

Solving the above full QP requires handling constraint(something like interior point method) but for unconstrained formulation(i.e ignoring A*X<=B) we can simply take the derivative which results in:

H*x+F=0 hence X=-H^-1 * F
Also, Autoware has a internal unconstrained solver similar to the above solution:

https://github.com/Autoware-AI/core_planning/blob/master/mpc_follower/src/qp_solver/qp_solver_unconstr.cpp